apis/v1beta1/gateway_types.go

```diff
@@ -0,0 +1,806 @@
+	// the ClientHello message of the TLS protocol.
+	//
+	// Note that SSL passthrough is only supported by TLSRoute.
+	TLSModePassthrough TLSModeType = "Passthrough"
+)
+
+// AllowedRoutes defines which Routes may be attached to this Listener.
+type AllowedRoutes struct {
+	// Namespaces indicates namespaces from which Routes may be attached to this
+	// Listener. This is restricted to the namespace of this Gateway by default.
+	//
+	// Support: Core
+	//
+	// +optional
+	// +kubebuilder:default={from: Same}
+	Namespaces *RouteNamespaces `json:"namespaces,omitempty"`
+
+	// Kinds specifies the groups and kinds of Routes that are allowed to bind
+	// to this Gateway Listener. When unspecified or empty, the kinds of Routes
+	// selected are determined using the Listener protocol.
+	//
+	// A RouteGroupKind MUST correspond to kinds of Routes that are compatible
+	// with the application protocol specified in the Listener's Protocol field.
+	// If an implementation does not support or recognize this resource type, it
+	// MUST set the "ResolvedRefs" condition to False for this Listener with the
+	// "InvalidRouteKinds" reason.
+	//
+	// Support: Core
+	//
+	// +optional
+	// +kubebuilder:validation:MaxItems=8
+	Kinds []RouteGroupKind `json:"kinds,omitempty"`
+}
+
+// FromNamespaces specifies namespace from which Routes may be attached to a
+// Gateway.
+//
+// +kubebuilder:validation:Enum=All;Selector;Same
+type FromNamespaces string
+
+const (
+	// Routes in all namespaces may be attached to this Gateway.
+	NamespacesFromAll FromNamespaces = "All"
+	// Only Routes in namespaces selected by the selector may be attached to
+	// this Gateway.
+	NamespacesFromSelector FromNamespaces = "Selector"
+	// Only Routes in the same namespace as the Gateway may be attached to this
+	// Gateway.
+	NamespacesFromSame FromNamespaces = "Same"
+)
+
+// RouteNamespaces indicate which namespaces Routes should be selected from.
+type RouteNamespaces struct {
+	// From indicates where Routes will be selected for this Gateway. Possible
+	// values are:
+	// * All: Routes in all namespaces may be used by this Gateway.
+	// * Selector: Routes in namespaces selected by the selector may be used by
+	//   this Gateway.
+	// * Same: Only Routes in the same namespace may be used by this Gateway.
+	//
+	// Support: Core
+	//
+	// +optional
+	// +kubebuilder:default=Same
+	From *FromNamespaces `json:"from,omitempty"`
+
+	// Selector must be specified when From is set to "Selector". In that case,
+	// only Routes in Namespaces matching this Selector will be selected by this
+	// Gateway. This field is ignored for other values of "From".
+	//
+	// Support: Core
+	//
+	// +optional
+	Selector *metav1.LabelSelector `json:"selector,omitempty"`
+}
+
+// RouteGroupKind indicates the group and kind of a Route resource.
+type RouteGroupKind struct {
+	// Group is the group of the Route.
+	//
+	// +optional
+	// +kubebuilder:default=gateway.networking.k8s.io
```

Comment:
The group is the domain part of the `apiVersion` field, `v1` is the version. We've chosen to leave out version here to make it so that users don't have to manually update versions as we move through them - on the assumption we'll move through versions much faster than upstream Kubernetes does.